### PR TITLE
chore: enable scorecard github action

### PR DESCRIPTION
**Pull request by joycebrum**
#### Description of Change

Regarding the issue #35709 

1. Create the `scorecard.yaml` which enable and configure the Scorecard Github Action to run: `on push in the main branch` and `once a week`. Any doubts about the configuration file feel free to ask.
2. Add the [badge](https://openssf.org/blog/2022/09/08/show-off-your-security-score-announcing-scorecards-badges/) to the README file.

Fixes: https://github.com/electron/electron/issues/35709

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes:  none

### CODE REVIEW

**Review by MarshallOfSound — CHANGES_REQUESTED** [review]
While I understand the intention of such a scorecard system I'm not sure this actually serves anyone any tangible benefit and the things on which we "lose points" are arbitrary and in some cases missing context.

image
In this case the "unreviewed commits" are automation commits kicking in as part of the release process.  If anything they can be trusted more than the reviewed commits 🤷 but with this score system it's impossible to convey that.

image
Idk what this one is about, but we definitely have an npm package `electron` 🤷 seems like a bug maybe?

image
Not sure why we're being dinged for only needing a single reviewer.  Very few people (by security design) have write access to `electron/electron`.  Requiring two or more (not sure what the requirement is there) to review every PR is wildly detached from the reality that open source projects operate in.  Surely "the point of compromise is one of 7 people" is dramatically better than "30 people now have write access but hey at least 2 of them need to look at every PR".

The only two things I can see in this report that I personally think are valid things for us to think about (and we are) are `Signed-Releases` and `Pinned-Dependencies`.

Given the limitations outlined above, my disagreement with some of the arbitrarily chosen numbers and the seemingly limited benefit to the project of including this Magic Number in our README I'm currently -1 on this addition.

**Review by nornagon — COMMENTED** [review]
Broadly LGTM, just a few questions! Thanks for this :)

I see at https://api.securityscorecards.dev/projects/github.com/electron/electron there's some info about the sorts of things this check will report. Is there a better dashboard or similar?

Looking through that report, I see a lot of issues that I'm not sure are really relevant for us, or at least which I don't fully understand!

- "GitHub code reviews found for 22 commits out of the last 30 -- score normalized to 7"
  - I believe the commits which don't have code reviews are version bump commits created by a bot, which we don't want to require code review for. Is there any way we can exclude such bots? (... or should we find a way to remove commit access for those bots? cc @MarshallOfSound)
- "Determines if the project has an OpenSSF (formerly CII) Best Practices Badge."
  - What does this mean?
- "Warn: no GitHub publishing workflow detected"
  - This check doesn't seem accurate for our project; our release process clearly produces downloadable products.
- "0 out of 5 artifacts are signed or have provenance"
  - I'm curious what the reasoning is behind signing release artifacts—we've done some investigations into reproducible builds and so on, but given all our releases are built in CI, I'm not sure it adds much. I guess if we put a private key somewhere in the CI secrets, this would at least prove that the release was made by a machine that had that key?
- "Warn: number of required reviewers is only 1 on branch 'main'"
  - This is a project governance decision. I'm not sure we want to change this. I'm a bit skeptical that requiring 2+ reviewers is a significant security improvement. Many large OSS projects don't require this (e.g. Chromium).

**Comment by nornagon** [comment]
ah somewhat collided with @MarshallOfSound.

File: .github/workflows/scorecards.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+jobs:
+  analysis:
+    name: Scorecards analysis
+    runs-on: ubuntu-latest
+    permissions:
+      # Needed to upload the results to code-scanning dashboard.
+      security-events: write
+      # Used to receive a badge.
+      id-token: write
```
[the remark this answers is not in the document]
Comment (joycebrum) [reply .github/workflows/scorecards.yml:21]: `security-events: write` is the permission to write the results in the code-scanning tab at the [security dashboard](https://github.com/electron/electron/security) with proposed solutions. See example bellow

Screenshot 2022-09-26 10 46 18

`id-token` is used for the [REST API](https://github.com/ossf/scorecard-action#rest-api)

**Comment by joycebrum** [comment]
> Broadly LGTM, just a few questions! Thanks for this :)
> 
> I see at https://api.securityscorecards.dev/projects/github.com/electron/electron there's some info about the sorts of things this check will report. Is there a better dashboard or similar?

This result will be shown in the [security dashboard](https://github.com/electron/electron/security) with proposed solutions. See example bellow

Screenshot 2022-09-26 10 46 18


> Looking through that report, I see a lot of issues that I'm not sure are really relevant for us, or at least which I don't fully understand!
> 
> * "GitHub code reviews found for 22 commits out of the last 30 -- score normalized to 7"
>   
>   * I believe the commits which don't have code reviews are version bump commits created by a bot, which we don't want to require code review for. Is there any way we can exclude such bots? (... or should we find a way to remove commit access for those bots? cc @MarshallOfSound)

That's an interesting topic to bring to attention. On the one hand, any bot can be compromised (its credentials can be compromised, for example), or considering that the commits are not being signed, an attacker could easily send a commit spoofing the bot. This means that the bot having a not supervised access to the main branch could potentially be a security risk. 

On the other hand, we understand that it can not be reasonable for this project to review every version bump made by this bot (which seems to happen every day). Thus, being able to exclude the bot from reviews would be the best possible outcome.

Considering it, I've oppened this topic for discussion on the Scorecard project https://github.com/ossf/scorecard/issues/2302.

> * "Determines if the project has an OpenSSF (formerly CII) Best Practices Badge."
>   
>   * What does this mean?
The [OpenSSF (formerly CII) Best Practices Badge](https://bestpractices.coreinfrastructure.org/) is a set of security-focus best development practice for open source projects. 

It has 3 tiers: passing, silver and gold. To earn the passing badge, the project MUST:

- publish the process for reporting vulnerabilities on the project site
- provide a working build system that can automatically rebuild the software from source code (where applicable)
- have a general policy that tests will be added to an automated test suite when major new functionality is added
- meet various cryptography criteria where applicable
- have at least one primary developer who knows how to design secure software
- have at least one primary developer who knows of common kinds of errors that lead to vulnerabilities in this kind of software (and at least one method to counter or mitigate each of them)
- apply at least one static code analysis tool (beyond compiler warnings and "safe" language modes) to any proposed major production release.
- Some of these criteria overlap with other Scorecards checks.

The remediation step for this check is to sign up for the [OpenSSF Best Practices program](https://bestpractices.coreinfrastructure.org/).

Pleas read [CII Best Practices](https://github.com/ossf/scorecard/blob/main/docs/checks.md#cii-best-practices) for more information.

> * "Warn: no GitHub publishing workflow detected"
>   
>   * This check doesn't seem accurate for our project; our release process clearly produces downloadable products.
> * "0 out of 5 artifacts are signed or have provenance"
>   
>   * I'm curious what the reasoning is behind signing release artifacts—we've done some investigations into reproducible builds and so on, but given all our releases are built in CI, I'm not sure it adds much. I guess if we put a private key somewhere in the CI secrets, this would at least prove that the release was made by a machine that had that key?

Another great and solid question @nornagon, thanks! As far as I've seen, the Scorecard team is working on improving the Signed-Release and Packaging checks (https://github.com/ossf/scorecard/issues/2166)

Please take a look at [Packaging Check](https://github.com/ossf/scorecard/blob/main/docs/checks.md#packaging) to a better understanding of what is the current posture and limitations of the Scorecard on packaging check and at [Signed Releases](https://github.com/ossf/scorecard/blob/main/docs/checks.md#signed-releases) and []() to see more about the Signed-Release Check and why is it important to sign releases.


> * "Warn: number of required reviewers is only 1 on branch 'main'"
>   
>   * This is a project governance decision. I'm not sure we want to change this. I'm a bit skeptical that requiring 2+ reviewers is a significant security improvement. Many large OSS projects don't require this (e.g. Chromium).

About this, you are right about having 2+ reviewers not being a significant security improvement, it is one of the highest level of branch protection check as you can see at https://github.com/ossf/scorecard/blob/main/docs/checks.md#branch-protection.

The lower levels like scoring at tier 3 already provides a significant branch protection level. Verifications greater than that would be helpful on improving the security posture but not that significant.

**Comment by joycebrum** [comment]
@MarshallOfSound thanks for sharing your feedbacks, here are some comments on it.

> While I understand the intention of such a scorecard system I'm not sure this actually serves anyone any tangible benefit and the things on which we "lose points" are arbitrary and in some cases missing context.

I understand your concern @MarshallOfSound, but just make it clear, the scorecards score is a tool to help you understand specific areas to improve in order to strengthen the security posture of your project, not to negatively scores you on what you are "not doing". That's not the idea at all. 

> In this case the "unreviewed commits" are automation commits kicking in as part of the release process. If anything they can be trusted more than the reviewed commits 🤷 but with this score system it's impossible to convey that.

I see it is similar to @nornagon question, which is a very relevant and solid question to raise. I've created the issue https://github.com/ossf/scorecard/issues/2302 to discuss this point but it is important to consider that not bot is 100% reliable and there is cons and pros on not reviewing their changes. 

> Idk what this one is about, but we definitely have an npm package electron 🤷 seems like a bug maybe?

The [Packaging Check](https://github.com/ossf/scorecard/blob/main/docs/checks.md#packaging) is currently limited to repositories hosted on GitHub, and does not support other source hosting repositories (i.e., Forges), but it is something that the scorecard team is already working on.

The check currently looks for [GitHub packaging workflows](https://docs.github.com/en/packages/learn-github-packages/publishing-a-package) and language-specific GitHub Actions that upload the package to a corresponding hub, e.g., [Npm](https://www.npmjs.com/). They plan to add better support to query package manager hubs directly in the future, e.g., for [Npm](https://www.npmjs.com/), [PyPi](https://pypi.org/).

Your question is a great feedback to the Scorecard team.

> Not sure why we're being dinged for only needing a single reviewer. Very few people (by security design) have write access to electron/electron. Requiring two or more (not sure what the requirement is there) to review every PR is wildly detached from the reality that open source projects operate in. Surely "the point of compromise is one of 7 people" is dramatically better than "30 people now have write access but hey at least 2 of them need to look at every PR".

I understand your point and really it is a very high branch protection level, you are now scoring 8/10 in branch protection, which is a excellent score considering the average of all open source projects!

> The only two things I can see in this report that I personally think are valid things for us to think about (and we are) are Signed-Releases and Pinned-Dependencies.

Great, those two very intersting points to working on regarding a better security posture. But besides those two, I would recomend to also look trought those too.

- [CII Best Practices](https://github.com/ossf/scorecard/blob/main/docs/checks.md#cii-best-practices), as mentioned earlier, it is a set of security-focus best development practice for open source projects.
- [Fuzzing](https://github.com/ossf/scorecard/blob/main/docs/checks.md#fuzzing): Fuzzing, or fuzz testing, is the practice of feeding unexpected or random data into a program to expose bugs. Regular fuzzing is important to detect vulnerabilities that may be exploited by others, especially since attackers can also use fuzzing to find the same flaws.
- [SAST](https://github.com/ossf/scorecard/blob/main/docs/checks.md#sast): SAST is testing run on source code before the application is run. Using SAST tools can prevent known classes of bugs from being inadvertently introduced in the codebase.

These ones I think would be intersting to work on to improve the overall security of the project. Of course, you can do it as you please, it is just a suggestion on what I think it can be relevant for the electron.

File: README.md [hunk 2]
```diff
@@ -3,6 +3,7 @@
 [![CircleCI Build Status](https://circleci.com/gh/electron/electron/tree/main.svg?style=shield)](https://circleci.com/gh/electron/electron/tree/main)
 [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/4lggi9dpjc1qob7k/branch/main?svg=true)](https://ci.appveyor.com/project/electron-bot/electron-ljo26/branch/main)
 [![Electron Discord Invite](https://img.shields.io/discord/745037351163527189?color=%237289DA&label=chat&logo=discord&logoColor=white)](https://discord.gg/electronjs)
+[![OpenSSF Scorecard](https://api.securityscorecards.dev/projects/github.com/electron/electron/badge)](https://api.securityscorecards.dev/projects/github.com/electron/electron)
```
[the remark this answers is not in the document]
Comment (joycebrum) [reply README.md]: Removed 👍🏼

**Comment by joycebrum** [comment]
I've noticed I forgot to anwer about why it is intersting to sign releases.

About what you've mentioned on adding the key to the CI, it would be a way to guarantee that the builds generated by it are reliable. This way, an attacker could not replace or simulate a compromised release because he/she would not have the key to do the signature.

**Comment by nornagon** [comment]
@joycebrum I think if an attacker has the ability to upload a release to GitHub, they could likely much more easily get access to our CI infrastructure, which would allow them to run any build they like through our CI and produced a signed build. So this seems dubiously useful to me.

**Comment by joycebrum** [comment]
@nornagon I undestand why do you think that because, in fact, if the CI is compromised, the releases it generates, even if signed, it will be also compromised.

The main benefit that the signing releases could bring for now it is the guarantee that a specific artifact was released by a source that you approve or you say it is reliable. 

Although, there are already moves to make it even more influent on the download process. The OpenSSF is working on [Implementing the signature verification with NPM packages](https://github.blog/2022-08-08-new-request-for-comments-on-improving-npm-security-with-sigstore-is-now-open/) what would allow a consumer to automatically verifiy if the package they are downloading was generated through a reliable builder and if it is correctly signed.

The Releases Signature already has some benefits and it is moving to a future with even more security benefits both for consumers and maintainers.

Anyhow, it is only one of the many checks that the Scorecards have, which you could choose to not adopt right now but at least knowing its pros and cons and what it is its objective in the close future

**Comment by azeemshaikh38** [comment]
> @joycebrum I think if an attacker has the ability to upload a release to GitHub, they could likely much more easily get access to our CI infrastructure, which would allow them to run any build they like through our CI and produced a signed build. So this seems dubiously useful to me.

Signed-Releases isn't specifically about an attacker uploading a release to GitHub. The check is really looking whether the project follows the best practice of signing (or generating some other provenance like SLSA attestations) for its software artifacts. That fact that we only look for GitHub releases is a shortcoming in our automation (hard to automatically detect different ways in which projects release their binaries). Hope that makes sense.

**Review by MarshallOfSound — APPROVED** [review]
Approved per slack discussion

**Pull request merged**